### PR TITLE
Read ref from GITHUB_HEAD_REF env variable if it's present

### PR DESCRIPTION
**Pull request by radoslaw-panuszewski**
When using `axion-release` together with [actions/checkout](https://github.com/actions/checkout), we needed to add the special configuration:
```
steps:
  - uses: actions/checkout@v4
    with:
      ref: ${{ github.head_ref }}
```
Otherwise, the action checks out repository in a detached-HEAD state and `axion-release` is not able to get the branch name.

Even though the branch ref is not checked out by default, Github provides the `GITHUB_HEAD_REF` which contains branch name that triggered the workflow. After this PR, `axion-release` will read this variable if available or fall back to git to preserve old behavior.

### CODE REVIEW

File: src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java [hunk 1]
```diff
@@ -333,37 +304,54 @@ private void assertPathExists(String path) {
...
-            String branchName = branchName();
-            return new ScmPosition(revision, branchName, isClean);
+            return revision;
         } catch (IOException e) {
             throw new ScmException(e);
         }
     }
 
+    private String branchName() {
+        return branchNameFromEnvVariable().orElseGet(this::branchNameFromGit);
```
Comment (bgalek) [inline src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java:327]: @radoslaw-panuszewski could you add docs entry about this happening?

**Review by bgalek — APPROVED** [review]

File: src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java [hunk 2]
```diff
@@ -333,37 +304,54 @@ private void assertPathExists(String path) {
...
+    }
+
+    /**
+     * @return head ref or source branch of the pull request in a workflow run
+     */
+    private Optional<String> branchNameFromEnvVariable() {
+        return Optional.ofNullable(System.getenv("GITHUB_HEAD_REF"));
```
Comment (bgalek) [inline src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java:334]: We need to handle multiple trigger types here (based on GITHUB_EVENT_NAME):
`push` -> GITHUB_REF_NAME
`pull-request`, `pull_request_target` -> GITHUB_HEAD_REF

We should only apply this logic where GITHUB_REF_TYPE is `branch` probably too.

Sources: 
https://docs.github.com/en/actions/learn-github-actions/variables
https://docs.github.com/en/actions/learn-github-actions/variables

**Review by bgalek — CHANGES_REQUESTED** [review]
We need to handle multiple trigger types here (based on GITHUB_EVENT_NAME):
push -> GITHUB_REF_NAME
pull-request, pull_request_target -> GITHUB_HEAD_REF

We should only apply this logic where GITHUB_REF_TYPE is branch probably too.

Sources:
https://docs.github.com/en/actions/learn-github-actions/variables
https://docs.github.com/en/actions/learn-github-actions/variables

**Comment by bgalek** [comment]
@radoslaw-panuszewski nice ;) 
It looks like workdlow_dispatch|call|run have GITHUB_HEAD_REF set and it cannot be committed since the UI requires you to select branch

**Comment by bgalek** [comment]
last thing: please update docs :)

File: src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java (hunk 1 above)
Comment (radoslaw-panuszewski) [reply src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java]: Done

**Comment by radoslaw-panuszewski** [comment]
We don't need to `GITHUB_REF_NAME` env variable for `push` events, because in case for those events, GitHub checks out either branch or tag that was pushed

**Review by bgalek — APPROVED** [review]

**Comment by bgalek** [comment]
thx, that looks very nice!

**Comment by Zordid** [comment]
I never used to do this ref thing in my GitHub actions... it just worked. Do I now have to change anything coming from 1.18.0?

**Comment by bgalek** [comment]
@Zordid nope, it should be 100% backward compatibile

**Pull request merged**